### PR TITLE
[ReadCacheFunctionalTests] Setup to run read cache functional tests on an already mounted directory

### PR DESCRIPTION
**Pull request by ashmeenkaur**
### Description
This PR includes 
* refactoring and setup to run read cache functional tests with mounted directory flag.
* Shell script to run read cache tests on an already mounted directory.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - Manually ran the tests on a GCE VM.
2. Unit tests - NA
3. Integration tests - Via KOKORO

### CODE REVIEW

File: tools/integration_tests/read_cache/disabled_cache_ttl_test.go [hunk 1]
```diff
@@ -75,6 +76,17 @@ func (s *disabledCacheTTLTest) TestReadAfterObjectUpdateIsCacheMiss(t *testing.T
...
+	// Create storage client before running tests.
+	closeStorageClient := createStorageClient(t, &ts.ctx, &ts.storageClient)
+
+	// Run tests for mounted directory if the flag is set.
+	if setup.AreBothMountedDirectoryAndTestBucketFlagsSet() {
+		test_setup.RunTests(t, ts)
+		return
+	}
+
+	defer closeStorageClient()
```
Comment (Tulsishah) [inline tools/integration_tests/read_cache/disabled_cache_ttl_test.go:89]: This should be at before if condition

**Review by Tulsishah — DISMISSED** [review]
Minor comment. PTAL

File: tools/integration_tests/run_tests_mounted_directory.sh [hunk 2]
```diff
@@ -304,3 +304,122 @@ echo "logging:
...
+
+  cleanup_test_environment
+  generate_config_file "$cache_size_mb" "$enable_range_read_cache" "$cache_ttl"
+}
+
+function cleanup_test_environment() {
+  # Clean up any pre-existing log files.
+  rm -r /tmp/gcsfuse_read_cache_test_logs
+  mkdir /tmp/gcsfuse_read_cache_test_logs
+   # Clean up cache directory.
```
Comment (raj-prince) [inline tools/integration_tests/run_tests_mounted_directory.sh:322]: nit: comment indentation.

File: tools/integration_tests/run_tests_mounted_directory.sh [hunk 3]
```diff
@@ -304,3 +304,122 @@ echo "logging:
...
+
+# cache-file-for-range-read:false read-cache test.
+read_cache_test_setup 50 false 3600
+test_cases=(
+  "TestCacheFileForRangeReadFalseTest/TestRangeReadsWithCacheMiss"
+  "TestCacheFileForRangeReadFalseTest/TestConcurrentReads_ReadIsTreatedNonSequentialAfterFileIsRemovedFromCache"
+)
+for test_case in "${test_cases[@]}"; do
+  gcsfuse --config-file=/tmp/gcsfuse_config.yaml "$TEST_BUCKET_NAME" "$MOUNT_DIR" > /dev/null
+  GODEBUG=asyncpreemptoff=1 go test ./tools/integration_tests/read_cache/... -p 1 --integrationTest -v --mountedDirectory="$MOUNT_DIR" --testbucket="$TEST_BUCKET_NAME" -run "$test_case"
```
Comment (raj-prince) [inline tools/integration_tests/run_tests_mounted_directory.sh:354]: I see the read cache test setup_test.go, if we pass mountedDirectory, it only runs the test for mounted Directory. In that case we don't use testbucket?

Comment (ashmeenkaur) [reply tools/integration_tests/run_tests_mounted_directory.sh:354]: Read cache functional tests uses storage client which requires the bucket name to fetch data from GCS. That is why we are passing both test bucket and mounted directory in these tests.

File: tools/integration_tests/run_tests_mounted_directory.sh [hunk 4]
```diff
@@ -304,3 +304,122 @@ echo "logging:
...
+# local modification test.
+read_cache_test_setup 9 false 3600
+test_case="TestLocalModificationTest/TestReadAfterLocalGCSFuseWriteIsCacheMiss"
+gcsfuse --config-file=/tmp/gcsfuse_config.yaml "$TEST_BUCKET_NAME" "$MOUNT_DIR" > /dev/null
+GODEBUG=asyncpreemptoff=1 go test ./tools/integration_tests/read_cache/... -p 1 --integrationTest -v --mountedDirectory="$MOUNT_DIR" --testbucket="$TEST_BUCKET_NAME" -run "$test_case"
+sudo umount "$MOUNT_DIR"
+cleanup_test_environment
+
+# range read read-cache test.
+read_cache_test_setup 500 false 3600
```
Comment (raj-prince) [inline tools/integration_tests/run_tests_mounted_directory.sh:385]: Value of enable_range_read_cache right?

Also, 
nit: comment upper case every place.

Comment (ashmeenkaur) [reply tools/integration_tests/run_tests_mounted_directory.sh:385]: the first parameter is cache size, second parameter is value of enable_range_read_cache and 3rd parameter is cache ttl in seconds. 
I've kept the value of enable_range_read_cache false here as these tests perform range reads on already cached files/chunks and are not dependent on the flag.

Comment (raj-prince) [reply tools/integration_tests/run_tests_mounted_directory.sh:385]: Should we also replicate the test for enable_range_read_cache for true?

File: tools/integration_tests/run_tests_mounted_directory.sh [hunk 5]
```diff
@@ -304,3 +304,122 @@ echo "logging:
...
+  "TestRangeReadTest/TestRangeReadsBeyondReadChunkSizeWithoutChunkDownloaded"
+)
+for test_case in "${test_cases[@]}"; do
+  gcsfuse --config-file=/tmp/gcsfuse_config.yaml "$TEST_BUCKET_NAME" "$MOUNT_DIR" > /dev/null
+  GODEBUG=asyncpreemptoff=1 go test ./tools/integration_tests/read_cache/... -p 1 --integrationTest -v --mountedDirectory="$MOUNT_DIR" --testbucket="$TEST_BUCKET_NAME" -run "$test_case"
+  sudo umount "$MOUNT_DIR"
+  cleanup_test_environment
+done
+
+# Read only mount read-cache test.
```
Comment (raj-prince) [inline tools/integration_tests/run_tests_mounted_directory.sh:398]: Here also, we have only considered enable_range_read as false, should we replicate it for true?

**Review by raj-prince — COMMENTED** [review]
Minor comments.

File: tools/integration_tests/run_tests_mounted_directory.sh (hunk 4 above)
Comment (raj-prince) [reply tools/integration_tests/run_tests_mounted_directory.sh:385]: Okay, got it.

File: tools/integration_tests/run_tests_mounted_directory.sh [hunk 6]
```diff
@@ -304,3 +304,134 @@ echo "logging:
...
+  "TestReadOnlyTest/TestReadFileRandomlyLargerThanCacheCapacity"
+  "TestReadOnlyTest/TestReadMultipleFilesMoreThanCacheLimit"
+  "TestReadOnlyTest/TestReadMultipleFilesWithinCacheLimit"
+)
+
+run_read_only_test_with_read_cache() {
+  read_cache_test_setup 9 $1 3600
+  gcsfuse --o=ro --config-file=/tmp/gcsfuse_config.yaml "$TEST_BUCKET_NAME" "$MOUNT_DIR" > /dev/null
+  GODEBUG=asyncpreemptoff=1 go test ./tools/integration_tests/read_cache/... -p 1 --integrationTest -v --mountedDirectory="$MOUNT_DIR" --testbucket="$TEST_BUCKET_NAME" -run "$2"
+  sudo umount "$MOUNT_DIR"
```
[the remark this answers is not in the document]
Comment (ashmeenkaur) [reply tools/integration_tests/run_tests_mounted_directory.sh:416]: `read_cache_test_setup` includes `cleanup_test_environment`, so I did not do it twice.

**Review by raj-prince — APPROVED** [review]

**Pull request merged**